File: apps/ios/AgentDeck/UI/ConversationMap.swift
import SwiftUI

/// The conversation's table of contents: one row per exchange — what was
/// asked, how it ended — each a jump into the timeline. A two-hundred-turn
/// chat has no scrollbar worth the name; this is the one it earns.
struct ConversationMapSheet: View {
    var markers: [ConversationMarker]
    var onPick: (String) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Conversation")
                .font(.system(size: 15, weight: .semibold))
                .foregroundStyle(Palette.text)
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 0) {
                    ForEach(markers) { marker in
                        Button {
                            onPick(marker.id)
                        } label: {
                            VStack(alignment: .leading, spacing: 2) {
                                HStack {
                                    Text(marker.prompt)
                                        .font(.system(size: 13, weight: .semibold))
                                        .foregroundStyle(Palette.text)
                                        .lineLimit(1)
                                    Spacer(minLength: 8)
                                    Text(Timestamps.freshness(marker.at))
                                        .font(.system(size: 11))
                                        .foregroundStyle(Palette.muted)
                                }
                                if let reply = marker.reply {
                                    Text(reply)
                                        .font(.system(size: 12))
                                        .foregroundStyle(Palette.muted)
                                        .lineLimit(1)
                                }
                            }
                            .padding(.horizontal, 6)
                            .padding(.vertical, 8)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.top, 20)
        .padding(.bottom, 12)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(Palette.surface)
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
    }
}

File: apps/android/mobile/src/main/java/dev/agentdeck/mobile/ConversationMap.kt
package dev.agentdeck.mobile

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.rounded.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.agentdeck.shared.*

/**
 * The conversation's table of contents: one row per exchange — what was
 * asked, how it ended — each a jump into the timeline. A two-hundred-turn
 * chat has no scrollbar worth the name; this is the one it earns.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ConversationMapSheet(
    markers: List<ConversationMarker>,
    onPick: (String) -> Unit,
    onDismiss: () -> Unit,
) {
    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = Surface) {
        Column(Modifier.fillMaxWidth().padding(start = 16.dp, end = 16.dp, bottom = 24.dp)) {
            Text(
                "Conversation",
                color = Text,
                fontSize = 15.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.padding(bottom = 10.dp),
            )
            LazyColumn(Modifier.heightIn(max = 560.dp)) {
                items(markers, key = { "marker:${it.id}" }) { marker ->
                    Column(
                        Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(9.dp))
                            .clickable { onPick(marker.id) }
                            .padding(horizontal = 6.dp, vertical = 8.dp),
                    ) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                marker.prompt,
                                color = Text,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.SemiBold,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                modifier = Modifier.weight(1f),
                            )
                            Spacer(Modifier.width(8.dp))
                            Text(formatMessageTime(marker.at), color = Muted, fontSize = 11.sp)
                        }
                        marker.reply?.let { reply ->
                            Text(
                                reply,
                                color = Muted,
                                fontSize = 12.sp,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                modifier = Modifier.padding(top = 2.dp),
                            )
                        }
                    }
                }
            }
        }
    }
}
